### PR TITLE
[MOD-13049] Do not rely on Shard Idx to decide where to send Command in Cluster mode

### PR DESCRIPTION
**Pull request by JoanFM**
## Describe the changes in the pull request

MRCommand was storing shardIdx as an offset to the shard array in topology to send commands (and to resend them in distributed commands where CURSORS are part of it). This was error prone in the presence of topology changes. 

This fix stores the ID of the node to make sure CURSOR communications reach the same instance all the time.

Consistency of the solution in the presence of topology changes is guaranteed by the ASM mechanism (with some limitations/corner cases)

---

> [!NOTE]
> Switches command routing to use shard node IDs rather than array indices, updating iterators, cursor mappings, barriers, topology handling, and tests with proper ownership/memory management.
> 
> - **Core routing/refactor**:
>   - `MRCommand`: replace integer `targetShard` with node ID `char*`; add `targetShardIdx` for response tracking; update copy/free/init logic.
>   - `MRCluster_SendCommand`/`MRCluster_GetConn`: use node ID directly; remove reliance on shard ordering (and shard sorting APIs/comments removed).
> - **Iterator/IO path**:
>   - Iterators now duplicate node IDs per shard, set `targetShardIdx`, and log with both; DEL transitions unchanged but use ID in logs.
>   - `ShardResponseBarrier_Notify` signature now `uint16_t shardIndex`; all callers updated (`rpnet`, `dist_utils`).
> - **Hybrid cursor mappings**:
>   - New `CursorMapping` struct (`targetShard` id, `targetShardIdx`, `cursorId`) and `CursorMapping_Release`.
>   - Proper ownership transfer/dup of `targetShard` to avoid leaks; early-bailout cleanup; RESP2/RESP3 handlers updated.
>   - RPNet start path uses mappings to create `_FT.CURSOR READ` per shard by node ID.
> - **Topology**:
>   - OSS/Enterprise topology builders no longer sort shards; order noted as arbitrary; clone logic unchanged.
> - **Coordinator utils**:
>   - `getCursorCommand` transfers shard ID ownership when rewriting; barrier notifications use `targetShardIdx`.
> - **Tests**:
>   - Adjusted C++ tests to expect `targetShard == nullptr`, verify copy/idx semantics.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit ecd5a0f71fda1f0bc1737059012ffb917b7507d6. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: src/coord/rmr/rmr.c [hunk 1]
```diff
@@ -629,23 +630,23 @@ void iterCursorMappingCb(void *p) {
...
-
   // Create FT.CURSOR READ commands for each mapping
   for (size_t i = 1; i < numShardsWithMapping; i++) {
     it->cbxs[i].it = it;
     it->cbxs[i].privateData = MRIteratorCallback_GetPrivateData(&it->cbxs[0]);
 
     it->cbxs[i].cmd = MRCommand_Copy(cmd);
 
-    it->cbxs[i].cmd.targetShard = vsimOrSearch->mappings[i].targetShard;
+    it->cbxs[i].cmd.targetShard = rm_strdup(vsimOrSearch->mappings[i].targetShard);
```
[the remark this answers is not in the document]
Comment (JoanFM) [reply src/coord/rmr/rmr.c]: mappings do not own targetShard, so each MRCOmmand should be freed independently

File: src/coord/rpnet.c [hunk 2]
```diff
@@ -131,19 +131,19 @@ void shardResponseBarrier_Init(void *ptr, MRIterator *it) {
...
 // Callback invoked by IO thread for each shard reply to accumulate totals
 // This function implements the ReplyNotifyCallback signature
-void shardResponseBarrier_Notify(int16_t shardId, long long totalResults, bool isError, void *privateData) {
+void shardResponseBarrier_Notify(uint16_t shardIndex, long long totalResults, bool isError, void *privateData) {
   ShardResponseBarrier *barrier = (ShardResponseBarrier *)privateData;
 
   // Validate shardId bounds
   size_t numShards = atomic_load(&barrier->numShards);
-  if (shardId < 0 || shardId >= (int16_t)numShards) {
+  if (shardIndex < 0 || shardIndex >= numShards) {
```
Comment (GuyAv46) [inline src/coord/rpnet.c]: `shardIndex < 0` is always false with unsigned variables

File: src/coord/rmr/command.h [hunk 3]
```diff
@@ -33,8 +33,13 @@ typedef struct {
   /* Slots info offset - 0 if not set (first argument is always the command) */
   uint32_t slotsInfoArgIndex;
 
-  /* if not NULL, this value indicate to which shard the command should be sent */
-  char *targetShard;
+  /* if not NULL, this value indicate to which shard the command should be sent.*/
+  const char *targetShard;
```
Comment (GuyAv46) [inline src/coord/rmr/command.h]: Making it `const` yields warnings when freeing it. Consider not having it const, or fix the warning in the `MRCommand` free function

File: src/coord/rmr/command.h [hunk 4]
```diff
@@ -33,8 +33,13 @@ typedef struct {
...
+
+  /* Index of the target shard in the cluster's shards array when command is created. Useful to keep track of responses.
+  Can't be used to know where to send the command, since the cluster's shards array can change. In case of new shards added, ASM should control not to trim
+  the slots from the old shard until all cursors are terminated (expected time limit for this). */
+  uint16_t targetShardIdx;
```
Comment (GuyAv46) [inline src/coord/rmr/command.h:42]: This is not really a command context anymore, and should be moved to the iterator context in the future

Comment (JoanFM) [reply src/coord/rmr/command.h:42]: let's keep this for the future

File: src/coord/hybrid/dist_hybrid.c [hunk 5]
```diff
@@ -440,6 +440,9 @@ static int HybridRequest_prepareForExecution(HybridRequest *hreq, RedisModuleCtx
 
 static void FreeCursorMappings(void *mappings) {
   CursorMappings *vsimOrSearch = (CursorMappings *)mappings;
+  for (size_t i = 0; i < array_len(vsimOrSearch->mappings); i++) {
+    CursorMapping_Release(&vsimOrSearch->mappings[i]);
+  }
   array_free(vsimOrSearch->mappings);
```
Comment (Itzikvaknin) [inline src/coord/hybrid/dist_hybrid.c:446]:
```suggestion
  array_free_ex(vsimOrSearch->mappings, CursorMapping_Release);
```

File: src/coord/hybrid/hybrid_cursor_mappings.c [hunk 6]
```diff
@@ -103,23 +120,35 @@ static void processHybridResp3(processCursorMappingCallbackContext *ctx, MRReply
...
+                  CursorMapping_Release(&vsimOrSearch->mappings[array_len(vsimOrSearch->mappings) - 1]);
+                }
                 array_pop(vsimOrSearch->mappings);
             }
             break;
         }
         mapping.cursorId = cid;
         CursorMappings *vsimOrSearch = StrongRef_Get(*mappings[i]);
         RS_ASSERT(vsimOrSearch);
+        if (i == 1) {
```
Comment (Itzikvaknin) [inline src/coord/hybrid/hybrid_cursor_mappings.c:140]:
```suggestion
        if (i == INTERNAL_HYBRID_RESP3_LENGTH - 1) {
```

Comment (JoanFM) [reply src/coord/hybrid/hybrid_cursor_mappings.c:140]: here i goes up to 2

File: src/coord/rmr/rmr.c [hunk 7]
```diff
@@ -577,21 +577,22 @@ void iterStartCb(void *p) {
...
 
-    it->cbxs[targetShard].privateData = MRIteratorCallback_GetPrivateData(&it->cbxs[0]);
+    it->cbxs[targetShardIdx].privateData = MRIteratorCallback_GetPrivateData(&it->cbxs[0]);
   }
 
-// Set the first command to target the first shard (while not having copied it)
-  targetShard = 0;
-  cmd->targetShard = targetShard;
-  MRCommand_SetSlotInfo(cmd, shards[targetShard].slotRanges);
+  // Set the first command to target the first shard (while not having copied it)
```
Comment (Itzikvaknin) [inline src/coord/rmr/rmr.c:591]: Lets move it above the loop here ☝️

File: src/coord/hybrid/hybrid_cursor_mappings.c (hunk 6 above)
Comment (Itzikvaknin) [reply src/coord/hybrid/hybrid_cursor_mappings.c:140]: Lets remove `processHybridResp2`, its dead code and causing confusions such as here

File: src/coord/hybrid/dist_hybrid.c (hunk 5 above)
Comment (JoanFM) [reply src/coord/hybrid/dist_hybrid.c:446]: Here to make it generic it would be weird, since here we pass reference, not the element by value.

File: src/coord/hybrid/hybrid_cursor_mappings.c [hunk 8]
```diff
@@ -90,6 +99,14 @@ static void processHybridResp2(processCursorMappingCallbackContext *ctx, MRReply
         }
 
         RS_ASSERT(vsimOrSearch);
+        if (i == INTERNAL_HYBRID_RESP2_LENGTH - 1) {
+            //Transferring ownership at the tail to avoid potential leak of cmd->targetShard on early bailout
+            mapping.targetShard = cmd->targetShard;
+            cmd->targetShard = NULL; // transfer ownership
+        } else {
+            mapping.targetShard = rm_strdup(cmd->targetShard);
+        }
```
[the remark this answers is not in the document]
Comment (JoanFM) [reply src/coord/hybrid/hybrid_cursor_mappings.c:106]: done

File: src/coord/hybrid/hybrid_cursor_mappings.c (hunk 6 above)
Comment (JoanFM) [reply src/coord/hybrid/hybrid_cursor_mappings.c:140]: u can remove in another PR , I do not want to remove extra unrelated things here

**Comment by JoanFM** [comment]
> LGTM, Im not that familiar with ASM, so lets wait at least for another approval

Has nothing to do with ASM however. Just adding another shard to the topology could affect here

File: src/coord/rmr/rmr.c [hunk 9]
```diff
@@ -577,8 +577,12 @@ void iterStartCb(void *p) {
 
   it->cbxs = rm_realloc(it->cbxs, numShards * sizeof(*it->cbxs));
   MRCommand *cmd = &it->cbxs->cmd;
-  size_t targetShardIdx;
-  for (targetShardIdx = 1; targetShardIdx < numShards; targetShardIdx++) {
+  // Set the first command to target the first shard (while not having copied it)
+  cmd->targetShard = rm_strdup(shards[0].node.id);
+  cmd->targetShardIdx = 0;
+  MRCommand_SetSlotInfo(cmd, shards[0].slotRanges);
```
Comment (GuyAv46) [inline src/coord/rmr/rmr.c]: Why did we move it to before the loop? We set it after we make all the clones (`MRCommand_Copy`) to avoid cloning the shard-specific info

File: src/coord/hybrid/hybrid_cursor_mappings.c [hunk 10]
```diff
@@ -76,6 +82,9 @@ static void processHybridResp2(processCursorMappingCallbackContext *ctx, MRReply
                 CursorMappings *vsim = StrongRef_Get(ctx->vsimMappings);
                 CursorMappings *search = StrongRef_Get(ctx->searchMappings);
                 while (array_len(vsim->mappings) > array_len(search->mappings)) {
+                    if (array_len(vsim->mappings) > 0) {
+                      CursorMapping_Release(&vsim->mappings[array_len(vsim->mappings) - 1]);
+                    }
                     array_pop(vsim->mappings);
```
Comment (GuyAv46) [inline src/coord/hybrid/hybrid_cursor_mappings.c]:
```suggestion
                    CursorMapping *cur = array_pop(vsim->mappings);
                    CursorMapping_Release(cur);
```

File: src/coord/hybrid/hybrid_cursor_mappings.c [hunk 11]
```diff
@@ -103,23 +120,35 @@ static void processHybridResp3(processCursorMappingCallbackContext *ctx, MRReply
...
         MRReply_ToInteger(cursorId, &cid);
         // Check for early bailout (Cursor ID 0 means no cursor was opened)
         if (cid == 0) {
             // Pop all mappings from previous subqueries
             for (int j = 0; j < i; j++) {
                 CursorMappings *vsimOrSearch = StrongRef_Get(*mappings[j]);
+                if (array_len(vsimOrSearch->mappings) > 0) {
+                  CursorMapping_Release(&vsimOrSearch->mappings[array_len(vsimOrSearch->mappings) - 1]);
+                }
                 array_pop(vsimOrSearch->mappings);
```
Comment (GuyAv46) [inline src/coord/hybrid/hybrid_cursor_mappings.c]:
```suggestion
                CursorMapping *cur = array_pop(vsimOrSearch->mappings);
                CursorMapping_Release(cur);
```

File: src/coord/rmr/rmr.c (hunk 9 above)
Comment (JoanFM) [reply src/coord/rmr/rmr.c]: comment from Itzik, I see now why it was at the end.

**Review by GuyAv46 — DISMISSED** [review]
LGTM

**Review by GuyAv46 — APPROVED** [review]
